### PR TITLE
Floating-point ops counting and reloading

### PR DESCRIPTION
**Pull request by TevenLeScao**
This PR adds flo-based logging of the validation loss. We count gigaflos to avoid overflowing, and because parameter numbers are also counted in billions. We don't count operations linked to the embeddings (cf. [Scaling Laws for Neural Language Models](https://arxiv.org/abs/2001.08361), section 2.1.) Those are stored in the `args` as `args.gigaflos_no_embeds`, and exported with them in the `state_dict` at checkpoint saving time. Currently, reloading a model trained before this PR assumes its flos are 0, which is similar behaviour to the `consumed_train_samples`.

### CODE REVIEW

File: megatron/data/gpt_dataset.py [hunk 1]
```diff
@@ -237,10 +237,10 @@ def _build_index_mappings(name, data_prefix, documents, sizes,
                 last_epoch_num_samples = num_samples - \
                                          num_samples_from_epochs_minus_one
                 assert last_epoch_num_samples >= 0, \
-                    'last epoch number of samples should be non-negative.'
+                    f'last epoch number of samples {last_epoch_num_samples} should be non-negative.'
                 num_samples_per_epoch = (tokens_per_epoch - 1) // seq_length
                 assert last_epoch_num_samples < (num_samples_per_epoch + 1), \
-                    'last epoch number of samples exceeded max value.'
+                    f'last epoch number of samples {last_epoch_num_samples} exceeded max value {num_samples_per_epoch + 1}.'
```
Comment (TevenLeScao) [inline megatron/data/gpt_dataset.py:243]: unrelated to the rest; was helpful to debug with very small datasets.

**Comment by TevenLeScao** [comment]
Note this integrates cherry-picked fixes from the main Megatron codebase, most notably changes to integrate deepspeed.

File: megatron/training.py [hunk 2]
```diff
@@ -663,20 +666,30 @@ def train(forward_step_func, model, optimizer, lr_scheduler,
     report_memory_flag = True
     while iteration < args.train_iters:
         update_num_microbatches(args.consumed_train_samples)
+        if args.deepspeed:
```
Comment (TevenLeScao) [inline megatron/training.py:669]: cherry-picked fix

File: megatron/training.py [hunk 3]
```diff
@@ -764,10 +777,17 @@ def evaluate(forward_step_func, data_iterator, model, verbose=False):
                     forward_backward_func = forward_backward_pipelining_without_interleaving
             else:
                 forward_backward_func = forward_backward_no_pipelining
-            loss_dicts = forward_backward_func(
-                forward_step_func, data_iterator, model, optimizer=None,
-                timers=None, forward_only=True)
-
+            
+            if args.deepspeed:
```
Comment (TevenLeScao) [inline megatron/training.py:781]: cherry-picked fix

File: megatron/training.py [hunk 4]
```diff
@@ -807,6 +827,9 @@ def evaluate_and_print_results(prefix, forward_step_func,
             writer.add_scalar('{} validation vs samples'.format(key),
                               total_loss_dict[key].item(),
                               args.consumed_train_samples)
+            writer.add_scalar('{} validation vs gigaflos (without embeddings)'.format(key),
+                              total_loss_dict[key].item(),
+                              args.gigaflos_no_embeds)
```
Comment (stas00) [inline megatron/training.py:832]: Please sync with https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/39 as it's now using groups.

@VictorSanh, perhaps you could merge #39 and then it'll be easier for Teven to adjust. Or the other way around.

File: megatron/utils.py [hunk 5]
```diff
@@ -204,11 +205,24 @@ def get_ltor_masks_and_position_ids(data,
...
 
-def get_parameters_in_billions(model):
+def non_embedding_params(module):
+    embedding_param_names = [
+        f"{name}.weight" for name, module_type in module.named_modules() if isinstance(module_type, nn.Embedding) or isinstance(module_type, VocabParallelEmbedding)
+    ]
+    non_embedding_parameters = [
+        parameter for name, parameter in module.named_parameters() if name not in embedding_param_names
+    ]
+    return sum(p.ds_numel if hasattr(p,'ds_id') else p.nelement() for p in non_embedding_parameters)
```
[the remark this answers is not in the document]
Comment (stas00) [reply megatron/utils.py:215]: this assumes deepspeed is used, but it'd break if it's not there, no? e.g. if one tries to use this repo on their home computer and they have no multiple gpus to setup PP, they may want not activate deepspeed. Unless perhaps using PP=1 will do the trick?

Still might be safer not to make an assumption in the long run. I think it'd be good to still be able to run Meg alone w/o this repo crashing.

 Thoughts?

File: megatron/utils.py [hunk 6]
```diff
@@ -204,11 +205,24 @@ def get_ltor_masks_and_position_ids(data,
...
+
+def get_parameters_in_billions(model, exclude_embeddings=False):
     gpus_per_model = torch.distributed.get_world_size(group=mpu.get_model_parallel_group())
 
-    approx_parameters_in_billions = sum([sum([p.ds_numel if hasattr(p,'ds_id') else  p.nelement() for p in model_module.parameters()])
-                                        for model_module in model])
+    if exclude_embeddings:
+        approx_parameters_in_billions = sum([non_embedding_params(model_module) for model_module in model])
+    else:
+        approx_parameters_in_billions = sum([sum([p.ds_numel if hasattr(p,'ds_id') else p.nelement() for p in model_module.parameters()])
```
[the remark this answers is not in the document]
Comment (stas00) [reply megatron/utils.py:224]: same commentary as above.

File: megatron/data/gpt_dataset.py (hunk 1 above)
Comment (thomasw21) [reply megatron/data/gpt_dataset.py:243]:
```suggestion
          assert last_epoch_num_samples <= num_samples_per_epoch, \
                    f'last epoch number of samples {last_epoch_num_samples} exceeded max value {num_samples_per_epoch}.'
```

File: megatron/training.py [hunk 7]
```diff
@@ -642,6 +644,7 @@ def save_checkpoint_and_time(iteration, model, optimizer, lr_scheduler):
 def train(forward_step_func, model, optimizer, lr_scheduler,
           train_data_iterator, valid_data_iterator):
     """Train the model function."""
+    print(sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad))
```
Comment (thomasw21) [inline megatron/training.py:647]:
```suggestion
    print(f"Number of trainable parameters: {sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad)}")
```

**Review by thomasw21 — COMMENTED** [review]
I'm not yet familiar with the codebase, so I'm not the best person to review so I'll let the others approve.

**Comment by stas00** [comment]
btw, this:
```
sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad)
```
will count tied variables multiple times, you probably want:
```
sum(dict((p.data_ptr(), p.numel()) for p in model.parameters()).values())
``` 
- after you adjust for `p.requires_grad`.

**Comment by stas00** [comment]
Also please make sure to rebase to include this fix if you're testing:
https://github.com/bigscience-workshop/Megatron-DeepSpeed/commit/42fe3b3986c5414bfeb7affafcb0f4b6615ad86c

File: megatron/training.py (hunk 4 above)
Comment (VictorSanh) [reply megatron/training.py:832]: i merged PR #39

File: megatron/training.py [hunk 8]
```diff
@@ -649,7 +649,7 @@ def save_checkpoint_and_time(iteration, model, optimizer, lr_scheduler):
 def train(forward_step_func, model, optimizer, lr_scheduler,
           train_data_iterator, valid_data_iterator):
     """Train the model function."""
-    print(sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad))
+    print(f"Number of trainable parameters: {sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad)}")
```
Comment (stas00) [inline megatron/training.py:652]: and this most likely needs to be fixed to remove double/triple-counting of tied vars:
https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/40#issuecomment-893159613

File: megatron/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stas00) [reply megatron/utils.py]: ah great! 

the proposed change was just a style adjustment.

File: megatron/utils.py [hunk 9]
```diff
@@ -205,14 +205,23 @@ def get_ltor_masks_and_position_ids(data,
     return attention_mask, loss_mask, position_ids
 
 
+def param_size(parameter):
+    return parameter.ds_numel if hasattr(parameter, 'ds_id') else parameter.nelement()
+
+
+def param_count_without_doubles(param_list):
+    return sum(dict((p.data_ptr(), param_size(p)) for p in param_list).values())
+    # sum(dict((p.data_ptr(), param_size(p)) for submodel in model for p in submodel.parameters()).values())
```
Comment (stas00) [inline megatron/utils.py:214]: this can be removed then

File: megatron/utils.py [hunk 10]
```diff
@@ -205,14 +205,23 @@ def get_ltor_masks_and_position_ids(data,
     return attention_mask, loss_mask, position_ids
 
 
+def param_size(parameter):
+    return parameter.ds_numel if hasattr(parameter, 'ds_id') else parameter.nelement()
+
+
+def param_count_without_doubles(param_list):
```
Comment (stas00) [inline megatron/utils.py:212]: since this can be triplicates, and more, perhaps a more telling name would be just: `unique_param_count`?

**Comment by TevenLeScao** [comment]
> btw, this:
> 
> ```
> sum(p.numel() for submodel in model for p in submodel.parameters() if p.requires_grad)
> ```
> 
> will count tied variables multiple times, you probably want:
> 
> ```
> sum(dict((p.data_ptr(), p.numel()) for p in model.parameters()).values())
> ```
> 
>     * after you adjust for `p.requires_grad`.

So originally I wasn't sure I was testing it right (sorry, took a bit more time than I would have liked) but now I am quite certain that they both return the same number, with duplicates. Do we have other leads on this?

**Comment by stas00** [comment]
I think you're correct, Teven, I experimented some more with it and it looks like in general case this is actually not needed.

I got the original at https://stackoverflow.com/a/62764464/9201239 and I'm pretty sure I used it because I was getting duplicate counts. But perhaps it was in a different scenario.

I think these have to be 2 tensors with shared `.strorage()` to need checking a unique `data_ptr`, however in most (all?) of the model layers we deal with the shared vars are just pointers to the same python variable. So since `model.parameters` creates a `set` based on tensor variables, it actually never returns any tied variables.

set creation:
https://github.com/pytorch/pytorch/blob/1f16c22dc8251f01627ee73ad1ef69bd18e51447/torch/nn/modules/module.py#L1488

For example here `lm_head` which is tied to the first embedding doesn't even show up as if it doesn't exist:
```
from transformers import T5ForConditionalGeneration;
m = T5ForConditionalGeneration.from_pretrained("t5-small"); 
for k,v in m.named_parameters():
    print(k)
```

**Comment by TevenLeScao** [comment]
Sorry for the misunderstanding, I currently find that both return the number with duplicates! For the same reasons you've laid, this does not correspond to what I thought the code would do, so I'm still investigating.

**Comment by stas00** [comment]
Oh, you're getting **wrong** numbers with both approaches.

Let's look at the code/data.

How many shared embedings do you have in the model, right? What is the code that does the sharing (or tieing)?

Let's perhaps see the dump of the model structure as well?

The things to consider

```
In [5]: import torch

In [6]: a = b = torch.ones(1)

In [7]: id(a)
Out[7]: 140596158463936

In [8]: id(b)
Out[8]: 140596158463936

In [11]: a.data_ptr()
Out[11]: 94903924655104

In [12]: b.data_ptr()
Out[12]: 94903924655104
```

So now you can look at the params that you think are shared and do the above and see which storage the python variables point to and which tensor data ptrs point to.

**Comment by TevenLeScao** [comment]
Hey, sorry, I looked at the code a bit and I don't think I currently have the bandwidth to adapt it this way. I'd really like this to be functional before we launch the next round of training; here are ways I think we can achieve that:
- Count parameters somewhere else, at the start of training (either with a formula or instantiating something like an HF model  in CPU RAM and counting there)
- Ask MLM/DS to do that; they seemed to have some leads from the Slack discussions ("The same bookkeeping is used to compute norms and other global values")
- Merge as-is and warn that this can only be used for non-embedding counts
What do you think?

**Comment by stas00** [comment]
As I mentioned in my last comment, I will need context to work with to understand the problem and to be able to reproduce it.

Once I finished sorting out the checkpoints I can work on this if you can explain to me how to reproduce the issue.

>     * Count parameters somewhere else, at the start of training (either with a formula or instantiating something like an HF model  in CPU RAM and counting there)

I don't think we have enough memory to load the whole model on cpu.

>     * Ask MLM/DS to do that; they seemed to have some leads from the Slack discussions ("The same bookkeeping is used to compute norms and other global values")

Doesn't hurt to ask

>     * Merge as-is and warn that this can only be used for non-embedding counts

warnings don't usually work IMHO, but if an incomplete solution works for your needs, and you need it now as-is, merge it and open another issue to track to bring this to completion.

**Comment by TevenLeScao** [comment]
I really haven't had time to come back to this; I'll merge now with a warning and open an issue.

**Comment by thomasw21** [comment]
Maybe unrelated, but I found this in DS repo:

https://github.com/microsoft/DeepSpeed/blob/big-science/deepspeed/runtime/pipe/engine.py#L117-L126

**Comment by stas00** [comment]
Hi Teven,

This warning isn't great:
```
        warnings.warn("Parameter count with the embeddings will be inaccurate with PP > 1, as the first and last stage hold several copies of the embeddings")
```

as it logs on each gpu and resulting in hundreds of these flooding the log file. could we please change this to run it only global rank 0 please?

Thank you!
